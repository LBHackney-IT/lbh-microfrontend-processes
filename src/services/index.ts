export { default as locale } from "./locale";
export * as processes from "./processes";
