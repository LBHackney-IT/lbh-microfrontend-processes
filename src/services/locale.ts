import reactHtmlParser from "react-html-parser";

import { AssetAddress } from "@mtfh/common/lib/api/asset/v1";

const locale = {
  backButton: "Back",
  cancel: "Cancel",
  selectProcessDescription: () =>
    reactHtmlParser(
      "<p>Start a process by selecting a process category below. <br /> Where a process is not yet supported in <strong>Manage My Home</strong> these links will open the appropriate interim Google form.</p>",
    ),
  errors: {
    unableToFetchRecord: "There was a problem retrieving the record",
    unableToFetchRecordDescription:
      "Please try again. If the issue persists, please contact support.",
    unableToFindProcess: "Invalid Process",
    unableToFindProcessDescription: "We have been unable to find a process of this name.",
  },
  title: "Processes",
  loadingText: "Submitting...",
  components: {
    startProcess: {
      buttonLabel: "Start process",
      thirdPartyHeading: "Sharing information with third parties",
    },
    entitySummary: {
      tenurePaymentRef: "Tenure payment ref",
      address: (assetAddress: AssetAddress): string => {
        const {
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,
          addressLine4,
          postCode,
        } = assetAddress;

        return [
          postPreamble,
          addressLine1,
          addressLine2,
          addressLine3,

          addressLine4,
          postCode,
        ]
          .filter((addressLine) => !!addressLine)
          .join(" ");
      },
    },
  },
};

export default locale;
