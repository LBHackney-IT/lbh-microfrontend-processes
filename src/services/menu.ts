import * as processes from "./processes";

interface MenuProps {
  label: string;
  link: string;
  processes?: { label: string; link: string; targetType: string }[];
}

const getProcessLink = (processName: string) => {
  const process = processes[processName];
  return {
    targetType: process.targetType,
    label: process.title,
    link: `/processes/${process.processName}/start`,
  };
};

const menu: MenuProps[] = [
  {
    label: "Changes to a tenancy",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSdgJ9DSgGI0Aj7GO1bzLbbrArPabjS8DQwmvwb9ltB-qYYESA/viewform",
    processes: [getProcessLink(processes.soletojoint.processName)],
  },
  {
    label: "Record a tenancy / audit homecheck",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfyXscVvgCBhbBBe7ADpV6cUqJoKrrTbU0YYRB_uXS2J4AHMQ/viewform",
  },
  {
    label: "Find legal referral response",
    link: "https://docs.google.com/spreadsheets/d/1fd6FwBXvBo2DmX8qZaXUr7hX9sE6Zr7HZsPlA_OO8Hs/edit#gid=1251146140",
  },
  {
    label: "Request a legal referral",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScpHsoJQgq112HIMbW06YAFpggijwQuODJM483HGbX_PMfCMw/viewform",
  },
  {
    label: "Record notice served",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfRSMia4FeT6WPjrmRlsLpOXyJ4tFs3Sv_O0toxUbv_FmIjKw/viewform",
  },
  {
    label: "Rehousing",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScPXSs4-xxTH2N5dUwtfmEl5mpu7kxZ6mHe86Qy_rlDNu-GbA/viewform",
  },
  {
    label: "Person",
    link: "https://docs.google.com/forms/d/e/1FAIpQLScxrWFLF7M7A3B1yzY-LZFXkV0P0qv6igcQ4wEgWHQk8DaUCw/viewform",
  },
  {
    label: "Report Japanese knotweed",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSceU2H-VWi2ewaEiOAkOWULVFU2DoPSen8SRPyJrmQtRpEhfQ/viewform?gxids=7628",
  },
  {
    label: "Report overgrown tree",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSfLmlLraDyOfqYhdjkOfkNWWlHUUep05G5n0GN-ewizN0Fbuw/viewform?gxids=7628",
  },
  {
    label: "Report overgrown garden",
    link: "https://docs.google.com/forms/d/e/1FAIpQLSc3mOUUEQhdRiDsS0k9XDWlfGJG0CEFAC2rnC1sYmM-idt-Nw/viewform?gxids=7628",
  },
];

export default menu;
