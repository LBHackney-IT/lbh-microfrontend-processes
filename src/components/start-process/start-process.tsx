import { Link as RouterLink, useHistory } from "react-router-dom";

import { Form, Formik } from "formik";
import * as Yup from "yup";

import { locale } from "../../services";
import { IStartProcess } from "../../types";

import { Button, Checkbox, Link } from "@mtfh/common/lib/components";

import "./styles.scss";

const { components } = locale;

interface StartProcessProps {
  process: IStartProcess;
  backLink: string;
  processName: string;
}

export const schema = Yup.object({
  condition: Yup.bool().oneOf([true]),
});

export type FormData = Yup.Asserts<typeof schema>;

export const StartProcess = ({ processName, process, backLink }: StartProcessProps) => {
  const history = useHistory();
  const { thirdPartyCondition, thirdPartyComponent, riskHeading, riskComponent } =
    process;
  const hasthirdPartyComponent = thirdPartyCondition && thirdPartyComponent;

  const renderComponent = (component) => {
    if (!component) return null;
    const Component = component;
    return <Component />;
  };

  return (
    <Formik<FormData>
      initialValues={{
        condition: !hasthirdPartyComponent,
      }}
      validateOnChange
      validateOnMount
      validateOnBlur={false}
      validationSchema={schema}
      onSubmit={() => {
        history.push(`/processes/${processName}/generated-id`);
      }}
    >
      {(properties) => (
        <Form>
          {hasthirdPartyComponent && (
            <>
              <h3>{components.startProcess.thirdPartyHeading}</h3>
              {renderComponent(thirdPartyComponent)}
              <Checkbox onChange={properties.handleChange} id="condition">
                {thirdPartyCondition}
              </Checkbox>
            </>
          )}
          {riskHeading && <h3>{riskHeading}</h3>}
          {renderComponent(riskComponent)}
          <div className="start-process__actions">
            <Button
              disabled={!properties.isValid}
              isLoading={properties.isSubmitting}
              loadingText={locale.loadingText}
              type="submit"
              className="start-process__start-button"
              variant="chevron"
            >
              {components.startProcess.buttonLabel}
            </Button>
            <Link className="start-process__cancel-link" as={RouterLink} to={backLink}>
              {locale.cancel}
            </Link>
          </div>
        </Form>
      )}
    </Formik>
  );
};
