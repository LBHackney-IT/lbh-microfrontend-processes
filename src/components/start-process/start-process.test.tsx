import { render } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";

import { locale, processes } from "../../services";
import { StartProcess } from "./start-process";

test("it enables form once checkbox is selected", async () => {
  render(
    <StartProcess
      process={processes.soleToJoint.startProcess}
      backLink="back-link"
      processName={processes.soleToJoint.urlPath}
    />,
  );

  const checkbox = await screen.findByLabelText(
    `${processes.soleToJoint.startProcess.thirdPartyCondition}`,
  );

  const startProcessButton = await screen.findByText(
    locale.components.startProcess.buttonLabel,
  );

  expect(startProcessButton).toBeDisabled();

  checkbox.click();

  await waitFor(() => {
    expect(startProcessButton).not.toBeDisabled();
    startProcessButton.click();
  });

  expect(window.location.pathname).toContain("/processes/sole-to-joint/generated-id");
});

test("Renders without third party content", async () => {
  const { thirdPartyComponent, thirdPartyCondition, ...rest } =
    processes.soleToJoint.startProcess;

  const { container } = render(
    <StartProcess
      process={rest}
      backLink="back-link"
      processName={processes.soleToJoint.urlPath}
    />,
  );

  await waitFor(() =>
    expect(
      screen.queryByText(locale.components.startProcess.thirdPartyHeading),
    ).not.toBeInTheDocument(),
  );

  expect(container).toMatchSnapshot();
});

test("Renders without risk content", async () => {
  const { riskComponent, riskHeading, ...rest } = processes.soleToJoint.startProcess;

  const { container } = render(
    <StartProcess
      process={rest}
      backLink="back-link"
      processName={processes.soleToJoint.urlPath}
    />,
  );
  await waitFor(() => expect(container).toMatchSnapshot());
});
