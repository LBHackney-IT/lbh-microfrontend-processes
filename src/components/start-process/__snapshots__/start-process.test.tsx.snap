// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without risk content 1`] = `
<div>
  <form
    action="#"
  >
    <h3>
      Sharing information with third parties
    </h3>
    <p
      class="lbh-body-m"
    >
      By starting this process the tenant and proposed tenant must agree that information relating to this application will be shared with third parties, for instance to perform a credit check.
    </p>
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="condition"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="condition"
      >
        I have explained to the tenant and proposed tenant that their information will be shared with third parties.
      </label>
    </div>
    <div
      class="start-process__actions"
    >
      <button
        class="govuk-button lbh-button lbh-button--chevron start-process__start-button"
        type="submit"
      >
        Start process
        <svg
          fill="none"
          height="22"
          viewBox="0 0 20 22"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M0 0H8.54573L20 10H11.4543L0 0Z"
            fill="white"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M0 22L8.54573 22L20 10H11.4543L0 22Z"
            fill="#96CCAE"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <a
        class="govuk-link lbh-link lbh-link--no-visited-state start-process__cancel-link"
        href="/back-link"
      >
        Cancel
      </a>
    </div>
  </form>
</div>
`;

exports[`Renders without third party content 1`] = `
<div>
  <form
    action="#"
  >
    <h3>
      Sole to joint tenure application risks
    </h3>
    <div>
      <p
        class="lbh-body-m"
      >
        A joint tenancy is when two or more adults (aged 18 or over) are named on the tenure agreement. Joint tenants have equal rights and responsibilities under the tenure agreement for the duration of the tenancy.
      </p>
      <p
        class="lbh-body-m"
      >
        Joint tenants are both responsible, together and individually, for keeping to these tenure conditions and payments to the London borough of Hackney. Applying for a joint tenancy is at the discretion of the Housing team.
      </p>
      <h4
        class="lbh-heading-h4"
      >
        Things to consider if joint tenants separate
      </h4>
      <p
        class="lbh-body-m"
      >
        If a joint tenants relationship ends, discussions should be made about what will happen to the tenure.
      </p>
      <p
        class="lbh-body-m"
      >
        The tenant and proposed tenant need to understand:
      </p>
      <ul
        class="lbh-list lbh-list--bullet"
      >
        <li>
          <p
            class="lbh-body-m"
          >
            When can they end a joint tenure
          </p>
        </li>
        <li>
          <p
            class="lbh-body-m"
          >
            What will happen if they end a joint tenure but one of the tenants remains at the property
          </p>
        </li>
      </ul>
      <p
        class="lbh-body-m"
      >
        This information is available on the
         
        <a
          class="govuk-link lbh-link"
          href="https://hackney.gov.uk/your-tenancy-agreement"
          rel="noopener noreferrer"
          target="_blank"
        >
          terms and conditions of the tenancy
        </a>
      </p>
    </div>
    <div
      class="start-process__actions"
    >
      <button
        class="govuk-button lbh-button lbh-button--chevron start-process__start-button"
        type="submit"
      >
        Start process
        <svg
          fill="none"
          height="22"
          viewBox="0 0 20 22"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M0 0H8.54573L20 10H11.4543L0 0Z"
            fill="white"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M0 22L8.54573 22L20 10H11.4543L0 22Z"
            fill="#96CCAE"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <a
        class="govuk-link lbh-link lbh-link--no-visited-state start-process__cancel-link"
        href="/back-link"
      >
        Cancel
      </a>
    </div>
  </form>
</div>
`;
